### PR TITLE
2 Input Reshape `ref` implementation

### PR DESCRIPTION
**Pull request by CharlieL7**
* Depends on #2190 
* This implements the two input version of ONNX Reshape where the second input is not constant.
* For dynamic batch this 2 input version will be simplified to the one input version after `simplify_dynamic_ops`
  * I still need to create that matcher 
* Note how the 2 input version takes an output_buffer: `reshape(input_tensor, output_buffer)`
  * This is intended to isolate taking output dimensions as input to the `allocate` operator.

### CODE REVIEW

**Review by TedThemistokleous — APPROVED** [review]
I think some of your changes will be moot once #2190 gets in but overall this makes sense.

Nice comment block in reshape.hpp. Shows your intent with this change well.

File: test/onnx/gen_onnx.py [hunk 1]
```diff
@@ -5793,6 +5793,24 @@ def reshape_non_standard_test():
     return ([trans, res], [x], [y])
 
 
+@onnx_test()
+def reshape_variable_input_test0():
```
Comment (umangyadav) [inline test/onnx/gen_onnx.py:5797]: Why is it named `variable` I don't see any `None` in dimensions.

File: src/include/migraphx/op/reshape.hpp [hunk 2]
```diff
@@ -215,32 +231,51 @@ struct reshape
...
-        return result;
+            visit_all(result, args[0])([&](auto output, auto input) {
+                std::copy(input.begin(), input.end(), output.begin());
+            });
+            return result;
+        }
+        else
+        {
+            // 2 arg
+            assert(args[0].get_shape().elements() == args[1].get_shape().elements());
```
Comment (umangyadav) [inline src/include/migraphx/op/reshape.hpp:273]: It should throw instead of assert. If elements do not match then it is a runtime error.

Comment (umangyadav) [reply src/include/migraphx/op/reshape.hpp:273]: can you add verify_onnx where it would throw?

File: test/onnx/gen_onnx.py (hunk 1 above)
Comment (CharlieL7) [reply test/onnx/gen_onnx.py:5797]: Means the second input is variable rather than constant (as in the other tests).

**Review by umangyadav — APPROVED** [review]

File: src/onnx/parse_spacetodepth.cpp
[the remark this answers is not in the document]
Comment (CharlieL7) [reply src/onnx/parse_spacetodepth.cpp:1]: These are changes from #2190

File: src/include/migraphx/op/reshape.hpp [hunk 3]
```diff
@@ -215,32 +231,56 @@ struct reshape
...
-        check_shapes{inputs, *this, true}.has(1);
+        check_shapes{inputs, *this, true}.has(1, 2);
 
         auto n_neg_dims = std::count(dims.begin(), dims.end(), -1);
         if(n_neg_dims > 1)
             MIGRAPHX_THROW("reshape: Dimensions for reshape can only have one -1 dim");
 
         auto s0 = inputs.front();
-        if(s0.dynamic())
+        if(inputs.size() == 1)
```
[the remark this answers is not in the document]
Comment (CharlieL7) [reply src/include/migraphx/op/reshape.hpp:241]: I don't understand your question in this context

[the remark this answers is not in the document]
Comment (bpickrel) [reply src/include/migraphx/op/reshape.hpp:241]: my mistake; I misread the condition

**Review by bpickrel — APPROVED** [review]

**Pull request merged**